### DISCUSSION TITLE
Generic extension method for generic type without duplicate parameters

### DISCUSSION THREAD

**Discussion by PMExtra** [dotnet/csharplang#4094] [General] [answered]
For simple, I provide a minimal example that disappear some code about async or external implementation
I have a generic type `IRepository<TEntity, TKey>`, I want make an extension method `Get<TDto>(id)` for it.
For now, I can only do that like below:
```csharp
public static class RepositoryExtensions
{
    public static TDto Get<TEntity, TKey, TDto>(this IRepository<TEntity, TKey> repository, TKey id)
    {
        return Map<TEntity, TDto>(repository.Get(id));
    }
}
```
Then, I can call it:
```csharp
IRepository<User, Guid> repository = ...;
var dto = repository.Get<User, Guid, UserDto>(id);
```
But it's not what I want. I want something like `repository.Get<UserDto>(id)`, because `TEntity` and `TKey` is already specified in `repository`. They should be inferred automatically.

It may related #2821

What do you think?

**Comment by PMExtra** [comment]
The `Map<TEntity, TDto>(TEntity source)` in above is also a problem.
Refer: https://github.com/AutoMapper/AutoMapper/blob/master/src/AutoMapper/Mapper.cs#L34-L38
The method 
L34: `TDestination Map<TDestination>(object source)` and the 
L38: `TDestination Map<TSource, TDestination>(TSource source)`, it's a similar case.

I don't think we should make the duplicate code such like this. Especially I can't use `as` or `is` to cast the argument if the parameter is a generic type.

**Comment by svick** [comment, marked as answer]
This is a duplicate of the championed proposal https://github.com/dotnet/csharplang/issues/1349.

**Reply by PMExtra** [reply]
I've read #1349 before post this.
#1349 provides only something like `repository.Get<TDto: UserDto>(id)` or `repository.Get<var, var, UserDto>(id)`,
But I want to make an extension method that can be used like `repository.Get<UserDto>(id)`.

**Reply by YairHalberstadt** [reply]
How will the compiler know which type parameter you're specifying?

**Reply by PMExtra** [reply]
@YairHalberstadt That's I want to discuss in this post. Just for a viable example, make a new keyword:
```csharp
public static TDto Get<implicit TEntity, implicit TKey, TDto>(this IRepository<TEntity, TKey> repository, TKey id)
```
The `implicit` means this generic argument should be inferred from argument instead specify when call.

**Reply by CyrusNajmabadi** [reply]
First, you should know we are interested in this space.  I would personally like it if we could have some sort of partial inference.  e.g. in your example, `TEntity` and `TKey` would be inferred, and hte user woudl jus need to supply the type for `DTO`.  However, this still is a dupe of #1349.  That particular issue tracks us wanting to do better here, but it's open ended in terms of what syntax we'll end up using to solve this problem.  

Thanks!

**Comment by pinkfloydx33** [comment]
I know you're requesting this for the broader case, but for your specific example if you're using c# 8 you could always pull DIMs into the mix:

```csharp
public interface IRepository<TEntity, TKey>
{
    TEntity Get(TKey id);
    TDto Get<TDto>(TKey id) {
        return Map<TEntity, TDto>(Get(Id));
    } 
}
```

Depending on what/where Map is (ie, if it's being called on something external to the instance/interface) you can combine the two approaches by delegating the DIM with one generic to the extension with three. You'd have to specify all the generics for the invocation, but you'd only need to do it in a single place.

The user would call the dim directly. The downside to this approach is it only works if you have a variable of the interface type (and not one of the concrete repository) and it adds a method to the interface surface. The plus side being that the Dim could be overridden ie. for custom mapping logic on a specific repository
